### PR TITLE
Add GridGraph

### PR DESCRIPTION
**Pull request by finnbuck**
Added two new operations, for constructing square and triangular [grid graphs](https://en.wikipedia.org/wiki/Lattice_graph). Also added tests for these functions.

### CODE REVIEW

**Review by samanthaharper — COMMENTED** [review]
Looks good, you just need documentation and to fix a few little things.

File: gap/examples.gd [hunk 1]
```diff
@@ -48,3 +48,6 @@ DeclareOperation("PetersenGraph", [IsFunction]);
 DeclareConstructor("GeneralisedPetersenGraphCons", [IsDigraph, IsInt, IsInt]);
 DeclareOperation("GeneralisedPetersenGraph", [IsInt, IsInt]);
 DeclareOperation("GeneralisedPetersenGraph", [IsFunction, IsInt, IsInt]);
+
+DeclareGlobalFunction("DIGRAPHS_SquareGridGraph");
```
[the remark this answers is not in the document]
Comment (finnbuck) [reply gap/examples.gd:52]: Yup, that's okay! I'll start getting them in the right format now :)

File: doc/examples.xml [hunk 2]
```diff
@@ -309,3 +309,74 @@ gap> GeneralisedPetersenGraph(IsMutableDigraph, 9, 4);
...
+  <Returns>A digraph.</Returns>
+  <Description>
+    If <A>n</A> and <A>k</A> are positive integers then this operation returns
+    the <E>Square Grid graph</E> of the dimensions <A>n</A> by <A>k</A>. <P/>
+
+    From
+    <URL>https://en.wikipedia.org/wiki/Lattice_graph</URL>:
+    <P/>
+
+    <Q>The square grid graph with n by k dimensions is a graph consisting
```
Comment (alex-konovalov) [inline doc/examples.xml:325]: Arguments should be enclosed in `<A>`

File: doc/examples.xml [hunk 3]
```diff
@@ -309,3 +309,74 @@ gap> GeneralisedPetersenGraph(IsMutableDigraph, 9, 4);
...
+  <Returns>A digraph.</Returns>
+  <Description>
+    If <A>n</A> and <A>k</A> are positive integers then this operation returns
+    a <E>Triangular Grid Graph</E> of the dimensions <A>n</A> by <A>k</A>. <P/>
+
+    From
+    <URL>https://en.wikipedia.org/wiki/Lattice_graph</URL>:
+    <P/>
+
+    <Q>The triangular grid graph with n by k dimensions is a graph consisting
```
Comment (alex-konovalov) [inline doc/examples.xml:360]: Please enclose arguments in `<A>`

File: gap/examples.gi [hunk 4]
```diff
@@ -368,5 +368,91 @@ InstallMethod(GeneralisedPetersenGraph, "for a function, integer, integer",
 [IsFunction, IsInt, IsInt],
 GeneralisedPetersenGraphCons);
 
-InstallMethod(GeneralisedPetersenGraph, "for integer, integer", [IsInt, IsInt],
+InstallMethod(GeneralisedPetersenGraph, "for integer, integer",
```
Comment (alex-konovalov) [inline gap/examples.gi:371]: "for two integers" would be a better way to formulate this.

Comment (finnbuck) [reply gap/examples.gi:371]: Yes, sorry about that. I was trying to fix some linting related to line length and accidentally changed part of the Petersen Graph code. I will change it back to the way it was!

File: doc/examples.xml [hunk 5]
```diff
@@ -310,6 +310,83 @@ gap> GeneralisedPetersenGraph(IsMutableDigraph, 9, 4);
 </ManSection>
 <#/GAPDoc>
 
+<#GAPDoc Label="SquareGridGraph">
+<ManSection>
+  <Oper Name="SquareGridGraph" Arg="[filt, ]n, k"/>
+  <Returns>A digraph.</Returns>
+  <Description>
+    If <A>n</A> and <A>k</A> are positive integers then this operation returns
+    the <E>Square Grid graph</E> of the dimensions <A>n</A> by <A>k</A>. <P/>
```
Comment (wilfwilson) [inline doc/examples.xml:319]:
```suggestion
    the <E>square grid graph</E> of dimension <A>n</A> by <A>k</A>. <P/>
```

File: gap/examples.gi [hunk 6]
```diff
@@ -371,6 +371,91 @@ GeneralisedPetersenGraphCons);
...
+    for j in [1 .. (n - 1)] do
+      DigraphAddEdge(D, [((i - 1) * n) + j, ((i - 1) * n) + j + 1]);
+    od;
+  od;
+  for i in [1 .. (k - 1)] do
+    for j in [1 .. n] do
+      DigraphAddEdge(D, [((i - 1) * n) + j, ((i - 1) * n) + j + n]);
+    od;
+  od;
+  return DigraphSymmetricClosure(D);
```
Comment (wilfwilson) [inline gap/examples.gi:392]: No offence to your code, but I think for maintainability, we should go with this simplification:
```suggestion
  local D1, D2;
  D1 := DigraphSymmetricClosure(ChainDigraph(IsMutableDigraph, n));
  D2 := DigraphSymmetricClosure(ChainDigraph(IsMutableDigraph, k));
  DigraphCartesianProduct(D1, D2);
  return D1;
```

File: gap/examples.gi [hunk 7]
```diff
@@ -371,6 +371,91 @@ GeneralisedPetersenGraphCons);
...
+end);
+
+InstallMethod(SquareGridGraphCons,
+"for IsImmutableDigraph and two positive integers",
+[IsImmutableDigraph, IsPosInt, IsPosInt],
+function(filt, n, k)
+  local D;
+  D := MakeImmutable(SquareGridGraphCons(IsMutableDigraph, n, k));
+  SetIsMultiDigraph(D, false);
+  SetIsSymmetricDigraph(D, true);
```
Comment (wilfwilson) [inline gap/examples.gi:402]:
```suggestion
  SetIsSymmetricDigraph(D, true);
  SetIsBipartiteDigraph(D, n > 1 or k > 1);
  SetIsPlanarDigraph(D, true);
  SetIsConnectedDigraph(D, true);
  SetDigraphHasLoops(D, false);
```

File: doc/examples.xml [hunk 8]
```diff
@@ -310,6 +310,83 @@ gap> GeneralisedPetersenGraph(IsMutableDigraph, 9, 4);
...
+    then the digraph being created will be mutable, if <A>filt</A> is <Ref
+    Filt="IsImmutableDigraph"/>, then the digraph will be immutable.
+    If the optional first argument <A>filt</A> is not present, then <Ref
+      Filt="IsImmutableDigraph"/> is used by default.<P/>
+    
+<Example><![CDATA[
+gap> SquareGridGraph(5, 5);
+<immutable symmetric digraph with 25 vertices, 80 edges>
+gap> SquareGridGraph(IsMutable, 5, 5);
+<mutable digraph with 25 vertices, 80 edges>
```
Comment (wilfwilson) [inline doc/examples.xml:344]:
```suggestion
<immutable bipartite symmetric digraph with bicomponent sizes 13 and 12>
gap> SquareGridGraph(IsMutable, 3, 4);
<mutable digraph with 12 vertices, 34 edges>
```

File: tst/standard/examples.tst [hunk 9]
```diff
@@ -226,6 +226,18 @@ Error, the arguments <n> and <k> must be non-negative integers,
 gap> JohnsonDigraph(IsMutableDigraph, 4, 2);
 <mutable digraph with 6 vertices, 24 edges>
 
+#  SquareGridGraph
+gap> SquareGridGraph(5, 5);
+<immutable symmetric digraph with 25 vertices, 80 edges>
+gap> SquareGridGraph(IsMutable, 5, 5);
+<mutable digraph with 25 vertices, 80 edges>
```
Comment (wilfwilson) [inline tst/standard/examples.tst:233]: A bit more variation is a good idea (and moreover, variation from the manual example):
```suggestion
gap> SquareGridGraph(7, 7);
<immutable bipartite symmetric digraph with bicomponent sizes 25 and 24>
gap> SquareGridGraph(IsMutable, 5, 3);
<mutable digraph with 15 vertices, 44 edges>
```

File: gap/examples.gi [hunk 10]
```diff
@@ -371,6 +371,91 @@ GeneralisedPetersenGraphCons);
...
+  for i in [1 .. k] do
+    for j in [1 .. (n - 1)] do
+      DigraphAddEdge(D, [((i - 1) * n) + j, ((i - 1) * n) + j + 1]);
+    od;
+  od;
+  for i in [1 .. (k - 1)] do
+    for j in [1 .. n] do
+      DigraphAddEdge(D, [((i - 1) * n) + j, ((i - 1) * n) + j + n]);
+    od;
+  od;
```
Comment (wilfwilson) [inline gap/examples.gi:431]: This is just the square grid graph, right? So let's not duplicate code, and just start with that!
```suggestion
  D := SquareGridGraph(IsMutableDigraph, n, k);
```

File: doc/examples.xml [hunk 11]
```diff
@@ -310,6 +310,83 @@ gap> GeneralisedPetersenGraph(IsMutableDigraph, 9, 4);
...
+</ManSection>
+<#/GAPDoc>
+  
+<#GAPDoc Label="TriangularGridGraph">
+<ManSection>
+  <Oper Name="TriangularGridGraph" Arg="[filt, ]n, k"/>
+  <Returns>A digraph.</Returns>
+  <Description>
+    If <A>n</A> and <A>k</A> are positive integers then this operation returns
+    a <E>Triangular Grid Graph</E> of the dimensions <A>n</A> by <A>k</A>. <P/>
```
Comment (wilfwilson) [inline doc/examples.xml:356]:
```suggestion
    a <E>triangular grid graph</E> of dimension <A>n</A> by <A>k</A>. <P/>
```

File: gap/examples.gi [hunk 12]
```diff
@@ -371,6 +371,91 @@ GeneralisedPetersenGraphCons);
...
+end);
+
+InstallMethod(TriangularGridGraphCons,
+"for IsImmutableDigraph and two positive integers",
+[IsImmutableDigraph, IsPosInt, IsPosInt],
+function(filt, n, k)
+  local D;
+  D := MakeImmutable(TriangularGridGraphCons(IsMutableDigraph, n, k));
+  SetIsMultiDigraph(D, false);
+  SetIsSymmetricDigraph(D, true);
```
Comment (wilfwilson) [inline gap/examples.gi:447]:
```suggestion
  SetIsSymmetricDigraph(D, true);
  SetIsBipartiteDigraph(D, n * k in Difference([n, k], [1]));
  SetIsPlanarDigraph(D, true);
  SetIsConnectedDigraph(D, true);
  SetDigraphHasLoops(D, false);
```

File: doc/examples.xml [hunk 13]
```diff
@@ -310,6 +310,83 @@ gap> GeneralisedPetersenGraph(IsMutableDigraph, 9, 4);
...
+
+    From
+    <URL>https://en.wikipedia.org/wiki/Lattice_graph</URL>:
+    <P/>
+
+    <Q>A triangular grid graph is a graph that corresponds to a triangular grid.</Q> 
+    <P/>
+    
+    The triangular grid graph with <A>n</A> by <A>k</A> dimensions is a graph consisting
```
Comment (wilfwilson) [inline doc/examples.xml:365]:
```suggestion
    The triangular grid graph with dimension <A>n</A> by <A>k</A>
    is a symmetric digraph consisting
```

**Review by wilfwilson — COMMENTED** [review]
I've made some suggestion, which I'll commit via the GitHub interface. I'm then going to make a few more changes locally and then push them to your branch.

**Review by wilfwilson — APPROVED** [review]
I've made some further changes and I think it's good to go. Have a look @finnbuck and if there's anything you want to complain about that I did, or any further changes you'd like to make, go ahead!

File: gap/examples.gi (hunk 6 above)
Comment (finnbuck) [reply gap/examples.gi]: Yeah no worries. I just learnt about the cartesian product of graphs while researching for the rook's graph and it definitely seems like a cleaner way of doing things!

File: gap/examples.gi (hunk 10 above)
Comment (finnbuck) [reply gap/examples.gi]: Ah right, that's a good point!

**Comment by finnbuck** [comment]
No complaints, thanks! The triangular grid graph really is just the square grid graph with extra edges so it makes complete sense for it to just call the square grid graph first...

**Comment by finnbuck** [comment]
Also, is there anything in particular that needs to go in the merge message when the branch is merged? By default it just says 'Add GridGraph'.

**Review by wilfwilson — APPROVED** [review]
No not really, I've just cleaned up the history and force-pushed to resolve the merge conflicts again (the branch couldn't be merged as it was) - sorry if I'm stepping on your toes.

I just called the one commit "Add SquareGridGraph and TriangularGridGraph (PR #416)" – you're still down in the git history as the author, don't worry 🙂 

I'll merge if the tests pass, once again.

**Comment by finnbuck** [comment]
Ok, that sounds great. Thanks again for walking me through all the steps!

**Comment by wilfwilson** [comment]
You're welcome!

**Pull request merged**